### DISCUSSION TITLE
Dynamic switching between fields in the same controlled form

### DISCUSSION THREAD

**Discussion by Stigjb** [react-hook-form/react-hook-form#7191] [Q&A] [answered]
I have a settings page where the fields are identical dropdowns with the same set of options. In addition, because there can be a lot of options, I'm using dependent dropdowns where first select controls which options are available the second one[^1]. I need to use `react-select` as a controlled component, and the design calls for a single form visible on the page instead of a list of settings. Therefore, I want there to be only one form on the page, and that the user should use buttons to change which field it controls.

However, my current approach isn't working as expected. I have taken inspiration from the [Tab form example](https://react-hook-form.com/faqs#Howtoworkwithmodalortabforms) from the FAQ and added a data store with `useState`. The values in the store seem to be updated correctly when the form is submitted, but the values in the `Select`s don't stay in sync. I thought the controlled inputs would always be updated when the value changes because I'm using `watch`.

Here is a minimal replication in Code Sandbox: <https://codesandbox.io/s/wizardly-wildflower-ml8ii?file=/src/App.tsx>. By playing around, you will quickly run into inconsistencies where the category selector for instance displays "Bird" and the species selector a non-bird like "Cod".

Is there a way to make this work? I have tried hacking around with `useRef`s and `useEffect`s, but not been able to solve the issue completely, and in all cases ended up with messy code.

[^1]: Similar to what's described in [this issue](https://github.com/react-hook-form/react-hook-form/discussions/2508)

**Comment by eirikeis** [comment, marked as answer]
Hi Stig,

It looks like field.value is using the getValue function which does not trigger a rerender. You can overwrite this behaviour doing as follows:
```
render={({ field: { value: _, ...field } }) => (
    <ReactSelect
      id="species"
      options={options}
      value={watch(field.name)}
      {...field}
    />
)}
```

There might be a more sophisticated way, but this looks to be working. Here is a working sandbox:

[https://codesandbox.io/s/dynamic-switching-with-react-hook-form-and-react-select-ulver](url)

**Reply by Stigjb** [reply]
Thank you, this seems to do the trick! I found that excluding the `value` field from `fields` is redundant as long as the value is passed after the spread fields, and this feels a little cleaner to me.

```jsx
render={({ field }) => (
  <ReactSelect
    id="species"
    options={options}
    {...field}
    value={watch(field.name)}
  />
)}
```

**Reply by bluebill1049** [reply]
thank you, @eirikeis

**Reply by ShahriarKh** [reply]
Unfortunately, it's not working correctly.
